In [7]:
import krippendorff
import numpy as np
import pandas as pd
import re

In [ ]:
train_file_path = f'D:/UU/Sem3/SDM/social_data_mining/data/train.xlsx'
excel_data = pd.ExcelFile(train_file_path)
sheet_names = excel_data.sheet_names
print("Sheet names: ", sheet_names)

Sheet names:  ['INSTRUCTIONS', 'CODER1', 'CODER2', 'CODER3']


In [ ]:
train_coder1_df = pd.read_excel(train_file_path, sheet_name=sheet_names[1])
train_coder2_df = pd.read_excel(train_file_path, sheet_name=sheet_names[2])
train_coder3_df = pd.read_excel(train_file_path, sheet_name=sheet_names[3])

In [17]:
train_coder1_df.head()

ID  CODE                                               TEXT
0  672233899615850496  -1.0  Piece for @TheWorldPost: 2 wars connected by c...
1  672231727952072704   0.0  Soils and #Ocean Omitted From Paris #COP21 Age...
2  672231250329919488   0.0  To reverse the Ephedrine Disinformation,@UN mu...
3  672229503696019456   0.0  See photos &amp; highlights from #COP21 side e...
4  672229362708561921   0.0  Our own Ben Parr writes for SBS News on India'...

In [18]:
train_coder2_df.head()

ID  CODE                                               TEXT
0  672233899615850496  -1.0  Piece for @TheWorldPost: 2 wars connected by c...
1  672231727952072704  -1.0  Soils and #Ocean Omitted From Paris #COP21 Age...
2  672231250329919488   1.0  To reverse the Ephedrine Disinformation,@UN mu...
3  672229503696019456   1.0  See photos &amp; highlights from #COP21 side e...
4  672229362708561921   0.0  Our own Ben Parr writes for SBS News on India'...

In [19]:
train_coder3_df.head()

ID  CODE                                               TEXT
0  672233899615850496   0.0  Piece for @TheWorldPost: 2 wars connected by c...
1  672231727952072704   0.0  Soils and #Ocean Omitted From Paris #COP21 Age...
2  672231250329919488   1.0  To reverse the Ephedrine Disinformation,@UN mu...
3  672229503696019456   0.0  See photos &amp; highlights from #COP21 side e...
4  672229362708561921   1.0  Our own Ben Parr writes for SBS News on India'...

In [27]:
train_merge_1_2_df = train_coder1_df.merge(train_coder2_df, on='ID', suffixes=("_1", "_2"))
train_merge_1_2_df = train_merge_1_2_df.drop(columns=["TEXT_2"]).rename(columns={"TEXT_1": "TEXT"})
train_merge_1_2_df.head()

ID  CODE_1  \
0  672233899615850496    -1.0   
1  672231727952072704     0.0   
2  672231250329919488     0.0   
3  672229503696019456     0.0   
4  672229362708561921     0.0   

                                                TEXT  CODE_2  
0  Piece for @TheWorldPost: 2 wars connected by c...    -1.0  
1  Soils and #Ocean Omitted From Paris #COP21 Age...    -1.0  
2  To reverse the Ephedrine Disinformation,@UN mu...     1.0  
3  See photos &amp; highlights from #COP21 side e...     1.0  
4  Our own Ben Parr writes for SBS News on India'...     0.0

In [28]:
train_coder3_df_copy = train_coder3_df.copy().drop(columns=["TEXT"]).rename(columns={"CODE": "CODE_3"})
train_merge_1_2_3_df = train_merge_1_2_df.merge(train_coder3_df_copy, on='ID')
train_merge_1_2_3_df.head()

ID  CODE_1  \
0  672233899615850496    -1.0   
1  672231727952072704     0.0   
2  672231250329919488     0.0   
3  672229503696019456     0.0   
4  672229362708561921     0.0   

                                                TEXT  CODE_2  CODE_3  
0  Piece for @TheWorldPost: 2 wars connected by c...    -1.0     0.0  
1  Soils and #Ocean Omitted From Paris #COP21 Age...    -1.0     0.0  
2  To reverse the Ephedrine Disinformation,@UN mu...     1.0     1.0  
3  See photos &amp; highlights from #COP21 side e...     1.0     0.0  
4  Our own Ben Parr writes for SBS News on India'...     0.0     1.0

In [20]:
test_file_path = f'D:/UU/Sem3/SDM/social_data_mining/data/test.xlsx'
test_excel_data = pd.ExcelFile(test_file_path)
sheet_names = test_excel_data.sheet_names
print("Sheet names: ", sheet_names)

Sheet names:  ['INSTRUCTIONS', 'CODER1', 'CODER2', 'CODER3']


In [21]:
test_coder1_df = pd.read_excel(test_file_path, sheet_name=sheet_names[1])
test_coder2_df = pd.read_excel(test_file_path, sheet_name=sheet_names[2])
test_coder3_df = pd.read_excel(test_file_path, sheet_name=sheet_names[3])

In [22]:
test_coder1_df.head()

ID  CODE                                               TEXT
0  675826762543140864   1.0  @AmedeusKizito I massive milestone in helping ...
1  675826724605730816   1.0  https://t.co/hZwEOmVXpW  Time now for Ireland ...
2  675826603977515008   1.0  Very proud of my country today #France #Succes...
3  675826598772248576  -1.0  Brain farts of an angry and an ignorant man ht...
4  675826488755617793   1.0  I am honestly looking forward to spending a fe...

In [23]:
test_coder2_df.head()

ID  CODE                                               TEXT
0  675826762543140864   1.0  @AmedeusKizito I massive milestone in helping ...
1  675826724605730816   1.0  https://t.co/hZwEOmVXpW  Time now for Ireland ...
2  675826603977515008   1.0  Very proud of my country today #France #Succes...
3  675826598772248576  -1.0  Brain farts of an angry and an ignorant man ht...
4  675826488755617793   0.0  I am honestly looking forward to spending a fe...

In [24]:
test_coder3_df.head()

ID  CODE                                               TEXT
0  675826762543140864   1.0  @AmedeusKizito I massive milestone in helping ...
1  675826724605730816   1.0  https://t.co/hZwEOmVXpW  Time now for Ireland ...
2  675826603977515008   1.0  Very proud of my country today #France #Succes...
3  675826598772248576  -1.0  Brain farts of an angry and an ignorant man ht...
4  675826488755617793   1.0  I am honestly looking forward to spending a fe...

In [15]:
dict_path = f"D:/UU/Sem3/SDM/social_data_mining/data/COPSentimentDict.xlsx"
dict_df = pd.read_excel(dict_path)
cop_dict = dict_df.to_dict(orient="records")
cop_dict 

[{'TERM': 'acting', 'SENTIMENT': 0.0},
 {'TERM': 'action', 'SENTIMENT': 1.0},
 {'TERM': 'advice', 'SENTIMENT': 0.0},
 {'TERM': 'Agreement', 'SENTIMENT': 1.0},
 {'TERM': 'artwork', 'SENTIMENT': 1.0},
 {'TERM': 'assaults', 'SENTIMENT': -1.0},
 {'TERM': 'background', 'SENTIMENT': 1.0},
 {'TERM': 'battles', 'SENTIMENT': -1.0},
 {'TERM': 'believe', 'SENTIMENT': 0.0},
 {'TERM': 'benefits', 'SENTIMENT': 1.0},
 {'TERM': 'bogeyman', 'SENTIMENT': 1.0},
 {'TERM': 'brandalism', 'SENTIMENT': 1.0},
 {'TERM': 'breakdown', 'SENTIMENT': 1.0},
 {'TERM': 'breakthrough', 'SENTIMENT': 1.0},
 {'TERM': 'bring', 'SENTIMENT': 0.0},
 {'TERM': 'broken', 'SENTIMENT': -1.0},
 {'TERM': 'caring', 'SENTIMENT': 1.0},
 {'TERM': 'Chemtrails', 'SENTIMENT': -1.0},
 {'TERM': 'circulate', 'SENTIMENT': 0.0},
 {'TERM': 'clean', 'SENTIMENT': -1.0},
 {'TERM': 'climbing', 'SENTIMENT': -1.0},
 {'TERM': 'coal', 'SENTIMENT': -1.0},
 {'TERM': 'community', 'SENTIMENT': 1.0},
 {'TERM': 'confident', 'SENTIMENT': 0.0},
 {'TERM': 'conser

### Krippendorf

In [29]:
annotations = train_merge_1_2_3_df[['CODE_1', 'CODE_2', 'CODE_3']].to_numpy()
annotations

array([[-1., -1.,  0.],
       [ 0., -1.,  0.],
       [ 0.,  1.,  1.],
       ...,
       [nan,  0.,  0.],
       [nan,  0.,  0.],
       [nan, -1.,  1.]])

In [31]:
alpha = krippendorff.alpha(annotations, level_of_measurement='nominal')
print("Inter-annotator agreement: ", alpha)

Inter-annotator agreement:  0.0012539556100379379


## Part-4

### Task 4.1

Our different types of actors are Actors who are "Private Individuals" and "Business Actors"

In [2]:
from text_mining.data import get_research_dataframe

from sdm.config import get_db_connection

In [3]:
conn = get_db_connection(db_path="D:/UU/Sem3/SDM/social_data_mining/data/twitter.db")
df = get_research_dataframe(db=conn)
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26019 entries, 0 to 26018
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   account_id    26019 non-null  object
 1   tweet_text    26019 non-null  object
 2   account_type  26019 non-null  object
 3   lang          26019 non-null  object
 4   stance        25480 non-null  object
dtypes: object(5)
memory usage: 1016.5+ KB
None


account_id                                         tweet_text  \
0   77254498  "@ComposerCorner I am proud to be a #climatech...   
1   77254498  #ActionDay ~ Blog ~ Greener Green Gifts  ~ htt...   
2   77254498  #Agnotology ~ @Ecojustice files complaint w/Co...   
3   77254498  #AsPredicted ~ @Canada wins #FossiloftheDay ~ ...   
4   77254498  #AusPol ~ #COP21 ~ #MentalHealth of #Australia...   

          account_type lang stance  
0  Private individuals   en    For  
1  Private individuals   en    For  
2  Private individuals   en    For  
3  Private individuals   en    For  
4  Private individuals   en    For

### Task 4.2

In [8]:
from text_mining.utils import preprocess_tweets
from sklearn.feature_extraction.text import CountVectorizer
from collections import Counter
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
import re

In [9]:
def remove_urls(txt):
    return re.sub(r'https://t\.co/\w+', '', txt)

In [10]:
def preprocess_text(text):
    text = text.lower()  # Convert to lowercase
    text = re.sub(r'[^\w\s]', '', text)  # Remove punctuation
    text = re.sub(r'\d+', '', text)  # Remove numbers
    words = text.split()  # Tokenize
    words = [w for w in words if w not in ENGLISH_STOP_WORDS and len(w) > 2 and 'http' not in w]  # Remove stop words
    return words

In [11]:
df["tweet_text_wo_url"] = df["tweet_text"].apply(remove_urls)
df['text'] = df['tweet_text_wo_url'].apply(preprocess_text)
df.head(10)

account_id                                         tweet_text  \
0   77254498  "@ComposerCorner I am proud to be a #climatech...   
1   77254498  #ActionDay ~ Blog ~ Greener Green Gifts  ~ htt...   
2   77254498  #Agnotology ~ @Ecojustice files complaint w/Co...   
3   77254498  #AsPredicted ~ @Canada wins #FossiloftheDay ~ ...   
4   77254498  #AusPol ~ #COP21 ~ #MentalHealth of #Australia...   
5   77254498  #CES2016 ~  Less-Smart Cars? https://t.co/KXUa...   
6   77254498  #CES2016 ~  Less-Smart Cars? https://t.co/KXUa...   
7   77254498  #COP21 #Music ~ Be an ~ Earthizen ~ c/o @MrChu...   
8   77254498     #COP21 ..... ..... ...... yawn ...... #CdnPoli   
9   77254498  #COP21 Blog ~ Dear Journalism, ~ https://t.co/...   

          account_type lang stance  \
0  Private individuals   en    For   
1  Private individuals   en    For   
2  Private individuals   en    For   
3  Private individuals   en    For   
4  Private individuals   en    For   
5  Private individuals   en    For   
6  Private individuals   en    For   
7  Private individuals   en    For   
8  Private individuals   en    For   
9  Private individuals   en    For   

                                   tweet_text_wo_url  \
0  "@ComposerCorner I am proud to be a #climatech...   
1  #ActionDay ~ Blog ~ Greener Green Gifts  ~  #L...   
2  #Agnotology ~ @Ecojustice files complaint w/Co...   
3  #AsPredicted ~ @Canada wins #FossiloftheDay ~ ...   
4  #AusPol ~ #COP21 ~ #MentalHealth of #Australia...   
5  #CES2016 ~  Less-Smart Cars?  #Design #Securit...   
6  #CES2016 ~  Less-Smart Cars?  … #Design #Secur...   
7  #COP21 #Music ~ Be an ~ Earthizen ~ c/o @MrChu...   
8     #COP21 ..... ..... ...... yawn ...... #CdnPoli   
9  #COP21 Blog ~ Dear Journalism, ~  #ClimateChan...   

                                                text  
0  [composercorner, proud, climatechange, denier,...  
1  [actionday, blog, greener, green, gifts, law, ...  
2  [agnotology, ecojustice, files, complaint, wco...  
3  [aspredicted, canada, wins, fossiloftheday, ge...  
4  [auspol, cop, mentalhealth, australian, wheat,...  
5  [ces, lesssmart, cars, design, security, cdnpo...  
6  [ces, lesssmart, cars, design, security, cdnpo...  
7  [cop, music, earthizen, mrchuckd, cdnpoli, gre...  
8                               [cop, yawn, cdnpoli]  
9  [cop, blog, dear, journalism, climatechange, c...

In [12]:
ba_tweets = df[df["account_type"] == "Business actors"]["text"]
pi_tweets = df[df["account_type"] == "Private individuals"]["text"]

ba_word_counts = Counter([word for tweet in ba_tweets for word in tweet])
pi_word_counts = Counter([word for tweet in pi_tweets for word in tweet])

# print(ba_word_counts)

We are going to create a -1 to 1 scale where -1 represents Private individuals and 1 denotes Business actors 

In [13]:
vocabulary = set(ba_word_counts.keys()).union(set(pi_word_counts.keys()))

word_scores = {}
total_ba_words = sum(ba_word_counts.values())
total_pi_words = sum(pi_word_counts.values())

for word in vocabulary:
    # ba_freq = ba_word_counts[word] if word in ba_word_counts and ba_word_counts[word] > 70 else 0
    # pi_freq = pi_word_counts[word] if word in pi_word_counts and pi_word_counts[word] > 70 else 0
    ba_freq = ba_word_counts.get(word, 0) / total_ba_words
    pi_freq = pi_word_counts.get(word, 0) / total_pi_words
    if ba_freq + pi_freq > 0:
        word_scores[word] = (ba_freq - pi_freq) / (ba_freq + pi_freq)
        
print(word_scores)

{'degc': 1.0, 'consciencegmestrallet': 1.0, 'andozile': -1.0, 'بالتغير': -1.0, 'master': -1.0, 'misunderstanding': -1.0, 'statoilasa': 1.0, 'preserving': 0.8001947626316609, 'capitales': -1.0, 'renewables': 0.2959484612240661, 'australiaumbrellastrong': -1.0, 'behavior': 1.0, 'hungerwfpertharin': -1.0, 'smithsonianmag': -1.0, 'iankingsky': -1.0, 'csrchat': -1.0, 'adhere': 1.0, 'rounds': 1.0, 'coffee': 0.6877853141888917, 'periodismo': -1.0, 'aggregatingbundling': 1.0, 'khanoisseur': -1.0, 'forge': -0.16226375302361412, 'earthtoparison': -1.0, 'adelphi_berlin': 1.0, 'strike': 0.2862112674387841, 'bloom_energy': 1.0, 'caringforclimate': 1.0, 'sarrêtera': -1.0, 'weatherrelated': -1.0, 'transforms': 0.09144583299126942, 'outer': -1.0, 'noyaux': 1.0, 'posición': -1.0, 'energía': -0.05209182129304122, 'torontos': -1.0, 'hire': 0.2862112674387841, 'nações': -1.0, 'unusual': 0.2862112674387841, 'journalistes': 0.2862112674387841, 'expression': -1.0, 'productive': 0.5655856094014733, 'cleaning'

In [14]:
n = 10

ba_top_words = sorted(word_scores.items(), key=lambda x: x[1], reverse=True)[:n]
pi_top_words = sorted(word_scores.items(), key=lambda x: x[1])[:n]

print("Top 10 words for Business actors:")
for word, score in ba_top_words:
    print(f"{word}: {score}")

print("\n---------------------------------\n")

print("\nTop 10 words for Private individuals:")
for word, score in pi_top_words:
    print(f"{word}: {score}")

Top 10 words for Business actors:
degc: 1.0
consciencegmestrallet: 1.0
statoilasa: 1.0
behavior: 1.0
adhere: 1.0
rounds: 1.0
aggregatingbundling: 1.0
adelphi_berlin: 1.0
bloom_energy: 1.0
caringforclimate: 1.0

---------------------------------


Top 10 words for Private individuals:
andozile: -1.0
بالتغير: -1.0
master: -1.0
misunderstanding: -1.0
capitales: -1.0
australiaumbrellastrong: -1.0
hungerwfpertharin: -1.0
smithsonianmag: -1.0
iankingsky: -1.0
csrchat: -1.0


In [15]:
word = "put"

print(ba_word_counts[word])
print(pi_word_counts[word])

print(max(ba_word_counts.values()))
print(max(pi_word_counts.values()))

0
0
9057
16685


In [73]:
# vectorizer = CountVectorizer(stop_words='english')
# X = vectorizer.fit_transform(df['text'])
# word_counts = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())
# word_counts["account_type"] = df["account_type"]

# word_counts.head()

In [74]:
# grouped_counts = word_counts.groupby("account_type").sum()
# total_counts = grouped_counts.sum(axis=0)
# print(total_counts)

In [75]:
# # Log-odds is used to score and rank words

# alpha = 1   # Smoothning parameter
# log_odds = np.log((grouped_counts + alpha) / (grouped_counts.sum(axis=1).values[:, None] - grouped_counts + alpha)) - \
#            np.log((total_counts + alpha) / (total_counts.sum() - total_counts + alpha))
           
# log_odds = pd.DataFrame(log_odds.T, index=vectorizer.get_feature_names_out(), columns=grouped_counts.index)

In [76]:
# log_odds.head()

In [77]:
# top_words = log_odds.apply(lambda x: x.nlargest(10).index, axis=0)
# print(top_words)

In [78]:
# total_count_dict = total_counts.to_dict()
# print(total_count_dict)

### Task 4.3

In [16]:
def classify_account(df, account_id, word_scores):
    account_score = 0
    account_tweets = df[df["account_id"] == account_id]
    
    if account_tweets.empty:
        return {"account_id": account_id, "account_type": "Unknown", "error": "Account not found"}
    
    for index, row in account_tweets.iterrows():
        words = row["text"]
        account_score += sum(word_scores.get(word, 0) for word in words)
        
    prediction = "Business actors" if account_score > 0 else "Private individuals"
    
    return {"account_id": account_id, "account_type": prediction, "account_score": account_score, "error": None}

### Check with Private Individuals and Business Actors

In [17]:
print(classify_account(df, "77254498", word_scores))
print(classify_account(df, "35495232", word_scores))
print(classify_account(df, "274652076", word_scores))
print(classify_account(df, "198573449", word_scores))

{'account_id': '77254498', 'account_type': 'Private individuals', 'account_score': -4760.581966043007, 'error': None}
{'account_id': '35495232', 'account_type': 'Private individuals', 'account_score': -33.68420804407124, 'error': None}
{'account_id': '274652076', 'account_type': 'Business actors', 'account_score': 149.0610988592079, 'error': None}
{'account_id': '198573449', 'account_type': 'Business actors', 'account_score': 178.44715569331785, 'error': None}


In [18]:
df[df["account_type"] == "Business actors"]

account_id                                         tweet_text  \
869    274652076  #AdaptationFund #COP21 side event Dec 4 1:15-2...   
870    274652076  #AdaptationFund Board Chair reports on Fund's ...   
871    274652076  #AdaptationFund active participant at #COP21 a...   
872    274652076  #AdaptationFund multimedia exhibit set up at #...   
873    274652076  #AdaptationFund pledges at #COP21 at nearly US...   
...          ...                                                ...   
25988  198573449  We must act w/ ambition @ #COP21. Our future —...   
25989  198573449  What happens to #energy production w/out suffi...   
25990  198573449  When #Lima residents say it never rains, they'...   
25991  198573449  Where #water and #climatechange meet [blog]: h...   
25992  198573449  Why waste #water when we can #reuse it? INFOGR...   

          account_type lang   stance  \
869    Business actors   en      For   
870    Business actors   en      For   
871    Business actors   en      For   
872    Business actors   en      For   
873    Business actors   en      For   
...                ...  ...      ...   
25988  Business actors   en  Unclear   
25989  Business actors   en  Unclear   
25990  Business actors   en  Unclear   
25991  Business actors   en  Unclear   
25992  Business actors   en  Unclear   

                                       tweet_text_wo_url  \
869    #AdaptationFund #COP21 side event Dec 4 1:15-2...   
870    #AdaptationFund Board Chair reports on Fund's ...   
871    #AdaptationFund active participant at #COP21 a...   
872    #AdaptationFund multimedia exhibit set up at #...   
873    #AdaptationFund pledges at #COP21 at nearly US...   
...                                                  ...   
25988  We must act w/ ambition @ #COP21. Our future —...   
25989  What happens to #energy production w/out suffi...   
25990  When #Lima residents say it never rains, they'...   
25991  Where #water and #climatechange meet [blog]:  ...   
25992  Why waste #water when we can #reuse it? INFOGR...   

                                                    text  
869    [adaptationfund, cop, event, dec, directaccess...  
870    [adaptationfund, board, chair, reports, funds,...  
871    [adaptationfund, active, participant, cop, cru...  
872    [adaptationfund, multimedia, exhibit, set, cop...  
873    [adaptationfund, pledges, cop, nearly, usmmore...  
...                                                  ...  
25988  [act, ambition, cop, future, amp, childrens, d...  
25989  [happens, energy, production, wout, sufficient...  
25990  [lima, residents, say, rains, theyre, far, wro...  
25991  [water, climatechange, meet, blog, cop, climat...  
25992  [waste, water, reuse, infographic, cop, thirst...  

[9156 rows x 7 columns]

In [19]:
df[df["account_type"] == "Private individuals"]

account_id                                         tweet_text  \
0       77254498  "@ComposerCorner I am proud to be a #climatech...   
1       77254498  #ActionDay ~ Blog ~ Greener Green Gifts  ~ htt...   
2       77254498  #Agnotology ~ @Ecojustice files complaint w/Co...   
3       77254498  #AsPredicted ~ @Canada wins #FossiloftheDay ~ ...   
4       77254498  #AusPol ~ #COP21 ~ #MentalHealth of #Australia...   
...          ...                                                ...   
26014   35495232  RT @WorldBank: #COP21 helped us build support ...   
26015   35495232  RT @bertodiplo: Met w/ mayor Biltaji before he...   
26016   35495232  RT @guardianeco: French president Francois Hol...   
26017   35495232  RT @thomsonreuters: Putting our planet on the ...   
26018   35495232  The Copenhagen-Paris express https://t.co/a8AZ...   

              account_type lang stance  \
0      Private individuals   en    For   
1      Private individuals   en    For   
2      Private individuals   en    For   
3      Private individuals   en    For   
4      Private individuals   en    For   
...                    ...  ...    ...   
26014  Private individuals   en    For   
26015  Private individuals   en    For   
26016  Private individuals   en    For   
26017  Private individuals   en    For   
26018  Private individuals   en    For   

                                       tweet_text_wo_url  \
0      "@ComposerCorner I am proud to be a #climatech...   
1      #ActionDay ~ Blog ~ Greener Green Gifts  ~  #L...   
2      #Agnotology ~ @Ecojustice files complaint w/Co...   
3      #AsPredicted ~ @Canada wins #FossiloftheDay ~ ...   
4      #AusPol ~ #COP21 ~ #MentalHealth of #Australia...   
...                                                  ...   
26014  RT @WorldBank: #COP21 helped us build support ...   
26015  RT @bertodiplo: Met w/ mayor Biltaji before he...   
26016  RT @guardianeco: French president Francois Hol...   
26017  RT @thomsonreuters: Putting our planet on the ...   
26018  The Copenhagen-Paris express  via @jordantimes...   

                                                    text  
0      [composercorner, proud, climatechange, denier,...  
1      [actionday, blog, greener, green, gifts, law, ...  
2      [agnotology, ecojustice, files, complaint, wco...  
3      [aspredicted, canada, wins, fossiloftheday, ge...  
4      [auspol, cop, mentalhealth, australian, wheat,...  
...                                                  ...  
26014  [worldbank, cop, helped, build, support, price...  
26015  [bertodiplo, met, mayor, biltaji, goes, cop, d...  
26016  [guardianeco, french, president, francois, hol...  
26017  [thomsonreuters, putting, planet, path, sustai...  
26018  [copenhagenparis, express, jordantimes, cop, p...  

[16863 rows x 7 columns]

### Check with accounts of different types

In [20]:
from text_mining.data import get_accounts_dataframe

accounts_df = get_accounts_dataframe(db=conn)
print(accounts_df.head())

  account_id                                         tweet_text  \
0  259352661  @DeutscheBank hat zw Gipfeln von Kopenhagen &a...   
1  259352661  @HanneloreKraft hat kurz vor #COP21 in #Kolumb...   
2  259352661  Activists' message to Paris ! #cop21 #quitcoal...   
3  259352661  New report! Big #banks financed $257B of #coal...   
4  259352661  RT @350: Lets be clear. If countries say they ...   

      account_type lang stance  
0  Advocacy actors   en    For  
1  Advocacy actors   en    For  
2  Advocacy actors   en    For  
3  Advocacy actors   en    For  
4  Advocacy actors   en    For  


In [24]:
accounts_df["tweet_text_wo_url"] = accounts_df["tweet_text"].apply(remove_urls)
accounts_df['text'] = accounts_df['tweet_text_wo_url'].apply(preprocess_text)

Test for Advocacy actor: 621925745; 259352661

In [25]:
accounts_df[accounts_df["account_type"] == "Advocacy actors"]

account_id                                         tweet_text  \
0      259352661  @DeutscheBank hat zw Gipfeln von Kopenhagen &a...   
1      259352661  @HanneloreKraft hat kurz vor #COP21 in #Kolumb...   
2      259352661  Activists' message to Paris ! #cop21 #quitcoal...   
3      259352661  New report! Big #banks financed $257B of #coal...   
4      259352661  RT @350: Lets be clear. If countries say they ...   
...          ...                                                ...   
84500  621925745  Un futuro de #HambreCero necesita resiliencia ...   
84501  621925745  Use #EarthToParis to tell world leaders a #Zer...   
84502  621925745  We're all benefit from a #ZeroHunger Planet. T...   
84503  621925745  Young farmers take ideals to Paris #COP21: htt...   
84504  621925745  “#Agriculture should be integrated in #climate...   

          account_type lang stance  \
0      Advocacy actors   en    For   
1      Advocacy actors   en    For   
2      Advocacy actors   en    For   
3      Advocacy actors   en    For   
4      Advocacy actors   en    For   
...                ...  ...    ...   
84500  Advocacy actors   en    For   
84501  Advocacy actors   en    For   
84502  Advocacy actors   en    For   
84503  Advocacy actors   en    For   
84504  Advocacy actors   en    For   

                                       tweet_text_wo_url  \
0      @DeutscheBank hat zw Gipfeln von Kopenhagen &a...   
1      @HanneloreKraft hat kurz vor #COP21 in #Kolumb...   
2      Activists' message to Paris ! #cop21 #quitcoal...   
3      New report! Big #banks financed $257B of #coal...   
4      RT @350: Lets be clear. If countries say they ...   
...                                                  ...   
84500  Un futuro de #HambreCero necesita resiliencia ...   
84501  Use #EarthToParis to tell world leaders a #Zer...   
84502  We're all benefit from a #ZeroHunger Planet. T...   
84503  Young farmers take ideals to Paris #COP21:  vi...   
84504  “#Agriculture should be integrated in #climate...   

                                                    text  
0      [deutschebank, hat, gipfeln, von, kopenhagen, ...  
1      [hannelorekraft, hat, kurz, vor, cop, kolumbie...  
2      [activists, message, paris, cop, quitcoal, par...  
3      [new, report, big, banks, financed, coal, cope...  
4      [lets, clear, countries, say, support, target,...  
...                                                  ...  
84500  [futuro, hambrecero, necesita, resiliencia, ca...  
84501  [use, earthtoparis, tell, world, leaders, zero...  
84502  [benefit, zerohunger, planet, tell, world, lea...  
84503  [young, farmers, ideals, paris, cop, stockandl...  
84504  [agriculture, integrated, climatechange, polic...  

[35619 rows x 7 columns]

In [36]:
print(classify_account(accounts_df, "621925745", word_scores))
print(classify_account(accounts_df, "259352661", word_scores))

{'account_id': '621925745', 'account_type': 'Private individuals', 'account_score': -197.8713583216119, 'error': None}
{'account_id': '259352661', 'account_type': 'Private individuals', 'account_score': -4.427085326307325, 'error': None}


Test for Political actors: 3232924232; 420704981

In [27]:
accounts_df[accounts_df["account_type"] == "Political actors"]

account_id                                         tweet_text  \
1940   3232924232  #COP21 is getting underway! 2 weeks of acceler...   
1941   3232924232  Adaptation is crucial in addressing #climatech...   
1942   3232924232  Africa shows the way at #COP21 by announcing 1...   
1943   3232924232  As we wait for the draft Paris Agreement, wort...   
1944   3232924232  Australian FM @JulieBishopMP: "The transformat...   
...           ...                                                ...   
84558   420704981  本日朝に新たな交渉文書が出るとのことなので、日本代表団は早めに会場入りしています。もうすぐ9...   
84559   420704981  来年COP22をホストするモロッコのパビリオンも輝いています！#COP21 https://...   
84560   420704981  気候世代エリアには、いたるところにハッシュタグが！#COP21 #GoCOP21 https...   
84561   420704981  現在フランスは午前1:33です。ハイレベルの会合は続いています。会議場に入れる人数は限られて...   
84562   420704981  ＊ECOなCOP21＊\n会場内には、このような分別ボックスがあちこちに置かれています。#C...   

           account_type lang   stance  \
1940   Political actors   en  Unclear   
1941   Political actors   en  Unclear   
1942   Political actors   en  Unclear   
1943   Political actors   en  Unclear   
1944   Political actors   en  Unclear   
...                 ...  ...      ...   
84558  Political actors   en  Unclear   
84559  Political actors   en  Unclear   
84560  Political actors   en  Unclear   
84561  Political actors   en  Unclear   
84562  Political actors   en  Unclear   

                                       tweet_text_wo_url  \
1940   #COP21 is getting underway! 2 weeks of acceler...   
1941   Adaptation is crucial in addressing #climatech...   
1942   Africa shows the way at #COP21 by announcing 1...   
1943   As we wait for the draft Paris Agreement, wort...   
1944   Australian FM @JulieBishopMP: "The transformat...   
...                                                  ...   
84558  本日朝に新たな交渉文書が出るとのことなので、日本代表団は早めに会場入りしています。もうすぐ9...   
84559             来年COP22をホストするモロッコのパビリオンも輝いています！#COP21    
84560          気候世代エリアには、いたるところにハッシュタグが！#COP21 #GoCOP21    
84561  現在フランスは午前1:33です。ハイレベルの会合は続いています。会議場に入れる人数は限られて...   
84562  ＊ECOなCOP21＊\n会場内には、このような分別ボックスがあちこちに置かれています。#C...   

                                                    text  
1940   [cop, getting, underway, weeks, accelerating, ...  
1941   [adaptation, crucial, addressing, climatechang...  
1942   [africa, shows, way, cop, announcing, renewabl...  
1943   [wait, draft, paris, agreement, worth, remembe...  
1944   [australian, juliebishopmp, transformation, gr...  
...                                                  ...  
84558  [本日朝に新たな交渉文書が出るとのことなので日本代表団は早めに会場入りしていますもうすぐ時で...  
84559                  [来年copをホストするモロッコのパビリオンも輝いていますcop]  
84560                [気候世代エリアにはいたるところにハッシュタグがcop, gocop]  
84561  [現在フランスは午前ですハイレベルの会合は続いています会議場に入れる人数は限られていますが会...  
84562        [ecoなcop, 会場内にはこのような分別ボックスがあちこちに置かれていますcop]  

[8902 rows x 7 columns]

In [35]:
print(classify_account(accounts_df, "3232924232", word_scores))
print(classify_account(accounts_df, "420704981", word_scores))

{'account_id': '3232924232', 'account_type': 'Business actors', 'account_score': 14.401266972842622, 'error': None}
{'account_id': '420704981', 'account_type': 'Private individuals', 'account_score': -0.2575627001299876, 'error': None}


Test for Journalistic actors: 2784989480; 22784404

In [29]:
accounts_df[accounts_df["account_type"] == "Journalistic actors"]

account_id                                         tweet_text  \
2577   2784989480  #ClimateChange:#Economic Collapse Of #Fishing/...   
2578   2784989480  @AdamSmithInt #ClimateChange:Collapse Of #Fish...   
2579   2784989480  @AfDB_Group #ClimateChange:Collapse Of #Fishin...   
2580   2784989480  @AidaOpokuMensah #ClimateChange:Collapse Of Fi...   
2581   2784989480  @Alllady247  #ClimateChange:Collapse Of #Fishi...   
...           ...                                                ...   
78391    22784404  RT @ACTAlliance: Good morning from ACT delegat...   
78392    22784404  RT @ACTAlliance: Mattias Södeberg talks about ...   
78393    22784404  Segunda pátria! #cop21 #therusnoinsta https://...   
78394    22784404  Stunning Paris #cop21 #nofilter https://t.co/D...   
78395    22784404  With my friends from Tanzania! #cop21 #therusn...   

              account_type lang   stance  \
2577   Journalistic actors   en  Unclear   
2578   Journalistic actors   en  Unclear   
2579   Journalistic actors   en  Unclear   
2580   Journalistic actors   en  Unclear   
2581   Journalistic actors   en  Unclear   
...                    ...  ...      ...   
78391  Journalistic actors   en  Unclear   
78392  Journalistic actors   en  Unclear   
78393  Journalistic actors   en  Unclear   
78394  Journalistic actors   en  Unclear   
78395  Journalistic actors   en  Unclear   

                                       tweet_text_wo_url  \
2577   #ClimateChange:#Economic Collapse Of #Fishing/...   
2578   @AdamSmithInt #ClimateChange:Collapse Of #Fish...   
2579   @AfDB_Group #ClimateChange:Collapse Of #Fishin...   
2580   @AidaOpokuMensah #ClimateChange:Collapse Of Fi...   
2581   @Alllady247  #ClimateChange:Collapse Of #Fishi...   
...                                                  ...   
78391  RT @ACTAlliance: Good morning from ACT delegat...   
78392  RT @ACTAlliance: Mattias Södeberg talks about ...   
78393             Segunda pátria! #cop21 #therusnoinsta    
78394                   Stunning Paris #cop21 #nofilter    
78395  With my friends from Tanzania! #cop21 #therusn...   

                                                    text  
2577   [climatechangeeconomic, collapse, fishingfarmi...  
2578   [adamsmithint, climatechangecollapse, fishingf...  
2579   [afdb_group, climatechangecollapse, fishingfar...  
2580   [aidaopokumensah, climatechangecollapse, fishi...  
2581   [alllady, climatechangecollapse, fishingfarmin...  
...                                                  ...  
78391  [actalliance, good, morning, act, delegation, ...  
78392  [actalliance, mattias, södeberg, talks, latest...  
78393              [segunda, pátria, cop, therusnoinsta]  
78394                   [stunning, paris, cop, nofilter]  
78395            [friends, tanzania, cop, therusnoinsta]  

[9653 rows x 7 columns]

In [34]:
print(classify_account(accounts_df, "2784989480", word_scores))
print(classify_account(accounts_df, "22784404", word_scores))

{'account_id': '2784989480', 'account_type': 'Business actors', 'account_score': 128.32234613598206, 'error': None}
{'account_id': '22784404', 'account_type': 'Private individuals', 'account_score': -8.854014304077717, 'error': None}


Test for Scientific actors: 43078165; 65339733

In [31]:
accounts_df[accounts_df["account_type"] == "Scientific actors"]

account_id                                         tweet_text  \
4169   462985721  RT @CECHR_UoD: LIVE #COP21 \nFrançois Hollande...   
9169    43078165  "Despite all these reasons for despair, I am h...   
9170    43078165  "It will be businesses, cities, states &amp; r...   
9171    43078165  "There is good news...political will is there ...   
9172    43078165  "We can't wait for the system to change throug...   
...          ...                                                ...   
81742   65339733  Using #trees + new varieties + water harvestin...   
81743   65339733  W Khaemba @ACTS: Ecosystem-based adaptation ne...   
81744   65339733  We won't understand global #climate change unt...   
81745   65339733  Yes, farmers are important players in the gree...   
81746   65339733  “Can our food system become part of the climat...   

            account_type lang stance  \
4169   Scientific actors   en    For   
9169   Scientific actors   en    For   
9170   Scientific actors   en    For   
9171   Scientific actors   en    For   
9172   Scientific actors   en    For   
...                  ...  ...    ...   
81742  Scientific actors   en    For   
81743  Scientific actors   en    For   
81744  Scientific actors   en    For   
81745  Scientific actors   en    For   
81746  Scientific actors   en    For   

                                       tweet_text_wo_url  \
4169   RT @CECHR_UoD: LIVE #COP21 \nFrançois Hollande...   
9169   "Despite all these reasons for despair, I am h...   
9170   "It will be businesses, cities, states &amp; r...   
9171   "There is good news...political will is there ...   
9172   "We can't wait for the system to change throug...   
...                                                  ...   
81742  Using #trees + new varieties + water harvestin...   
81743  W Khaemba @ACTS: Ecosystem-based adaptation ne...   
81744  We won't understand global #climate change unt...   
81745  Yes, farmers are important players in the gree...   
81746  “Can our food system become part of the climat...   

                                                    text  
4169   [cechr_uod, live, cop, françois, hollande, joi...  
9169           [despite, reasons, despair, hopeful, cop]  
9170   [businesses, cities, states, amp, regions, del...  
9171                 [good, newspolitical, parties, cop]  
9172   [wait, change, normal, cycles, change, climate...  
...                                                  ...  
81742  [using, trees, new, varieties, water, harvesti...  
81743  [khaemba, acts, ecosystembased, adaptation, ne...  
81744  [wont, understand, global, climate, change, af...  
81745  [yes, farmers, important, players, green, econ...  
81746          [food, climate, cure, cop, climatechange]  

[3364 rows x 7 columns]

In [33]:
print(classify_account(accounts_df, "43078165", word_scores))
print(classify_account(accounts_df, "65339733", word_scores))

{'account_id': '43078165', 'account_type': 'Business actors', 'account_score': 667.0476043567345, 'error': None}
{'account_id': '65339733', 'account_type': 'Business actors', 'account_score': 225.7276941556703, 'error': None}


### Testing with Stances

Test with Against Stance:  37271053; 21150492

In [37]:
accounts_df[accounts_df["stance"] == "Against"]

account_id                                         tweet_text  \
145     37271053  #COP21 \nDo it!\nDo it!\nDo it!\n\nDo the HUST...   
146     37271053  #COP21 \nDo it!\nDo it!\nDo it!\nDo the HUSTLE...   
147     37271053  #COP21  Activists Claim ‘Climate Hustle’ film ...   
148     37271053  #COP21  Data shows Greenland ice mass gain lin...   
149     37271053  #COP21  From the Mother Nature Climate Network...   
...          ...                                                ...   
65050   21150492  Wow... hours after #SanBernardino, dimwitted c...   
65051   21150492  Yeesh... French President Hollande just closed...   
65052   21150492  Yes, even the NYTimes says my anti-capitalist ...   
65053   21150492  Yes. Coal is the fastest growing source of ene...   
65054   21150492  You can't see him, but Obama is holding the 'N...   

              account_type lang   stance  \
145        Advocacy actors   en  Against   
146        Advocacy actors   en  Against   
147        Advocacy actors   en  Against   
148        Advocacy actors   en  Against   
149        Advocacy actors   en  Against   
...                    ...  ...      ...   
65050  Private individuals   en  Against   
65051  Private individuals   en  Against   
65052  Private individuals   en  Against   
65053  Private individuals   en  Against   
65054  Private individuals   en  Against   

                                       tweet_text_wo_url  \
145    #COP21 \nDo it!\nDo it!\nDo it!\n\nDo the HUST...   
146    #COP21 \nDo it!\nDo it!\nDo it!\nDo the HUSTLE...   
147    #COP21  Activists Claim ‘Climate Hustle’ film ...   
148    #COP21  Data shows Greenland ice mass gain lin...   
149    #COP21  From the Mother Nature Climate Network...   
...                                                  ...   
65050  Wow... hours after #SanBernardino, dimwitted c...   
65051  Yeesh... French President Hollande just closed...   
65052  Yes, even the NYTimes says my anti-capitalist ...   
65053  Yes. Coal is the fastest growing source of ene...   
65054  You can't see him, but Obama is holding the 'N...   

                                                    text  
145                                [cop, hustle, hustle]  
146                                [cop, hustle, hustle]  
147    [cop, activists, claim, climate, hustle, film,...  
148    [cop, data, shows, greenland, ice, mass, gain,...  
149    [cop, mother, nature, climate, network, look, ...  
...                                                  ...  
65050  [wow, hours, sanbernardino, dimwitted, climate...  
65051  [yeesh, french, president, hollande, just, clo...  
65052  [yes, nytimes, says, anticapitalist, jihad, dy...  
65053  [yes, coal, fastest, growing, source, energy, ...  
65054  [obama, holding, colonialism, sign, dineshdsou...  

[1840 rows x 7 columns]

In [38]:
print(classify_account(accounts_df, "37271053", word_scores))
print(classify_account(accounts_df, "21150492", word_scores))

{'account_id': '37271053', 'account_type': 'Private individuals', 'account_score': -391.4466869907296, 'error': None}
{'account_id': '21150492', 'account_type': 'Private individuals', 'account_score': -1339.4538954379404, 'error': None}


Test with For Stance: 108648075; 358393963; 172161666; 50257070; 96735431

In [47]:
accounts_df[accounts_df["account_id"] == "50257070"]

account_id                                         tweet_text  \
4978   50257070  Atlas Kasbah #Agadir #Maroc, une réponse local...   
4979   50257070  Live depuis la #COP21 avec @bertrandpiccard : ...   
4980   50257070  Live from @SolutionsCOP21 : l'excellent discou...   
4981   50257070  Live from @SolutionsCOP21 : très beau discours...   
4982   50257070  People's Climate March in #Agadir #Morocco #CO...   
4983   50257070  RT @Mlyhamza: Climat : #Maroc bon élève. La CO...   
4984   50257070  RT @Visit_Morocco_: Ils se sont engagés pour l...   
4985   50257070  Ravi que l'écolodge Atlas Kasbah représente #A...   
4986   50257070  Solutions #COP21 : Le #Maroc, un modèle de dév...   
4987   50257070  l'#ecolodge Atlas Kasbah d'#agadir #Maroc sur ...   
4988   50257070  l'ONMT se mobilise pour le climat = Faîtes vou...   

         account_type lang stance  \
4978  Business actors   en    For   
4979  Business actors   en    For   
4980  Business actors   en    For   
4981  Business actors   en    For   
4982  Business actors   en    For   
4983  Business actors   en    For   
4984  Business actors   en    For   
4985  Business actors   en    For   
4986  Business actors   en    For   
4987  Business actors   en    For   
4988  Business actors   en    For   

                                      tweet_text_wo_url  \
4978  Atlas Kasbah #Agadir #Maroc, une réponse local...   
4979  Live depuis la #COP21 avec @bertrandpiccard : ...   
4980  Live from @SolutionsCOP21 : l'excellent discou...   
4981  Live from @SolutionsCOP21 : très beau discours...   
4982  People's Climate March in #Agadir #Morocco #CO...   
4983  RT @Mlyhamza: Climat : #Maroc bon élève. La CO...   
4984  RT @Visit_Morocco_: Ils se sont engagés pour l...   
4985  Ravi que l'écolodge Atlas Kasbah représente #A...   
4986  Solutions #COP21 : Le #Maroc, un modèle de dév...   
4987  l'#ecolodge Atlas Kasbah d'#agadir #Maroc sur ...   
4988  l'ONMT se mobilise pour le climat = Faîtes vou...   

                                                   text  
4978  [atlas, kasbah, agadir, maroc, une, réponse, l...  
4979  [live, depuis, cop, avec, bertrandpiccard, hom...  
4980  [live, solutionscop, lexcellent, discours, dal...  
4981  [live, solutionscop, très, beau, discours, roy...  
4982    [peoples, climate, march, agadir, morocco, cop]  
4983  [mlyhamza, climat, maroc, bon, élève, cop, ser...  
4984  [visit_morocco_, ils, sont, engagés, pour, cli...  
4985  [ravi, que, lécolodge, atlas, kasbah, représen...  
4986  [solutions, cop, maroc, modèle, développement,...  
4987  [lecolodge, atlas, kasbah, dagadir, maroc, sur...  
4988  [lonmt, mobilise, pour, climat, faîtes, vous, ...

In [49]:
print(classify_account(accounts_df, "108648075", word_scores))
print(classify_account(accounts_df, "358393963", word_scores))
print(classify_account(accounts_df, "172161666", word_scores))
print(classify_account(accounts_df, "50257070", word_scores))
print(classify_account(accounts_df, "96735431", word_scores))

{'account_id': '108648075', 'account_type': 'Private individuals', 'account_score': -10.560345501121468, 'error': None}
{'account_id': '358393963', 'account_type': 'Private individuals', 'account_score': -11.741488986935154, 'error': None}
{'account_id': '172161666', 'account_type': 'Business actors', 'account_score': 487.07677415443106, 'error': None}
{'account_id': '50257070', 'account_type': 'Business actors', 'account_score': 44.84395055366322, 'error': None}
{'account_id': '96735431', 'account_type': 'Business actors', 'account_score': 97.6380619775439, 'error': None}


Above, tried to check if Stance has some meaningful association with Account type.

Found no such relationship, classification is meaningfully associated only with account type. Detailed patterns are explained below

## Conclusion

Our model is quite good at classifying text from Private Individuals and Business actors. 

Scores < 0 indicate Private Individual account type, while score > 0 indicates Business actor account type.

For example: These accounts were correctly classified.

{'account_id': '77254498', 'account_type': 'Private individuals', 'account_score': -4760.581966043007, 'error': None} 

{'account_id': '35495232', 'account_type': 'Private individuals', 'account_score': -33.68420804407124, 'error': None}

{'account_id': '274652076', 'account_type': 'Business actors', 'account_score': 149.0610988592079, 'error': None}

{'account_id': '198573449', 'account_type': 'Business actors', 'account_score': 178.44715569331785, 'error': None}

We also tried 2 cases each of the other account types:

**Advocacy actors:** Both of our advocacy actor example accounts were classified as Private individuals

{'account_id': '621925745', 'account_type': 'Private individuals', 'account_score': -197.8713583216119, 'error': None}

{'account_id': '259352661', 'account_type': 'Private individuals', 'account_score': -4.427085326307325, 'error': None}

**Political actors:** Our examples were randomly classified based on the words being used.

{'account_id': '3232924232', 'account_type': 'Business actors', 'account_score': 14.401266972842622, 'error': None}

{'account_id': '420704981', 'account_type': 'Private individuals', 'account_score': -0.2575627001299876, 'error': None}

**Journalistic actors:** Our examples were randomly classified based on the words being used.

{'account_id': '2784989480', 'account_type': 'Business actors', 'account_score': 128.32234613598206, 'error': None}

{'account_id': '22784404', 'account_type': 'Private individuals', 'account_score': -8.854014304077717, 'error': None}

**Scientific actors:** Both of our scientific actor example accounts were classified as Business actors

{'account_id': '43078165', 'account_type': 'Business actors', 'account_score': 667.0476043567345, 'error': None}

{'account_id': '65339733', 'account_type': 'Business actors', 'account_score': 225.7276941556703, 'error': None}